### PR TITLE
Change error store to be part of configuration instead

### LINKED ISSUE #27: Error Store should not require a storage spec
The error store should be defined on deployment (of core), and should not require a spec.

### PR DESCRIPTION
**Pull request by zhilingc**
Fixes #27.

This change requires core to be deployed with the following additional configuration:
- `feast.jobs.errorsStoreType`
- `feast.jobs.errorsStoreOptions`

which define where (and how) to write the error logs of each ingestion job. These options are passed in as flags to the ingestion jar when a job is run. Internally to ingestion a storage spec is still used.

### CODE REVIEW

File: ingestion/src/main/java/feast/ingestion/boot/ImportJobModule.java [hunk 1]
```diff
@@ -88,7 +98,7 @@ Specs provideSpecs(SpecService.Builder specService) {
 
   @Provides
   @Singleton
-  List<ErrorsStore> provideErrorsStores() {
-    return ErrorsStoreService.getAll();
+  ErrorsStore provideErrorsStores() {
```
Comment (tims) [inline ingestion/src/main/java/feast/ingestion/boot/ImportJobModule.java]: this should still return a list, as this is about returning the available classes in the class path, which are essential factories for building write transforms. The passed in errorsStoreType is used for selecting the write one from this list.

**Comment by tims** [comment]
We could change the [Error|Serving|Warehouse]Store classes to have factory in the name if that makes them make more sense. I just found it very verbose.

File: ingestion/src/main/java/feast/storage/service/ErrorsStoreService.java [hunk 2]
```diff
@@ -17,42 +17,18 @@
...
-      }
-    }
-    return null;
+  public static ErrorsStore get() {
+    return errorsStore;
   }
 
   public static void register(ErrorsStore store) {
-    manuallyRegistered.add(store);
+    errorsStore = store;
```
[the remark this answers is not in the document]
Comment (zhilingc) [reply ingestion/src/main/java/feast/storage/service/ErrorsStoreService.java]: @tims

[the remark this answers is not in the document]
Comment (tims) [reply ingestion/src/main/java/feast/storage/service/ErrorsStoreService.java]: No, they only get configured when you can create, and pass them a storage spec. This just makes some factory objects that get ignored.

File: ingestion/src/main/java/feast/ingestion/boot/ImportJobModule.java (hunk 1 above)
Comment (zhilingc) [reply ingestion/src/main/java/feast/ingestion/boot/ImportJobModule.java]: Made changes as suggested

**Comment by zhilingc** [comment]
@tims

File: ingestion/src/main/java/feast/storage/service/ErrorsStoreService.java [hunk 3]
```diff
@@ -38,7 +39,7 @@
 
   public static List<ErrorsStore> getAll() {
     return Lists.newArrayList(
-        Iterators.concat(manuallyRegistered.iterator(), serviceLoader.iterator()));
+            Iterators.concat(manuallyRegistered.iterator(), serviceLoader.iterator()));
```
Comment (tims) [inline ingestion/src/main/java/feast/storage/service/ErrorsStoreService.java]: This is weird, are you using the google-java-format plugin or the imported googles styles xml?
https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml

The imports don't get rearranged for me either.

**Comment by tims** [comment]
Looks good except for formatting and import org, which need the google styles applied.
i.e: https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing

If the google-java-format plugin neglects to org imports correctly. So instead, import the intellij styles.
https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml

File: ingestion/src/main/java/feast/ingestion/options/ImportJobOptions.java [hunk 4]
```diff
@@ -72,11 +72,19 @@
   void setLimit(Long value);
 
   @Description(
-      "Set a store id to store errors in, if your data input is **very** small, you can use STDOUT"
-          + " or STDERR as the store id, otherwise it must match an associated storage spec")
-  String getErrorsStoreId();
+      "Set an errors store type. One of: [STDERR, STDOUT, JSON]. Note that you should not use "
```
[the remark this answers is not in the document]
Comment (zhilingc) [reply ingestion/src/main/java/feast/ingestion/options/ImportJobOptions.java]: oh right, thanks :)

**Pull request merged**